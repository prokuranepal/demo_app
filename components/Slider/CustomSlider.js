import React, { useState } from 'react';
import { StyleSheet, View, Text } from 'react-native';
import Slider from './SliderContainer';;

const CustomSlider = props => {

  const defaultValue = 0
  const maximumValue = 3;
  const length = 4
  const sliderRatio = maximumValue / (length - 1);
  const [state, setState] = useState({
    value: props.initialValue,
    adjustSign: 1
  });
  console.log("state value",state.value)
  const sliderOptions = [
    { value: 0, label: 'Eat' },
    { value: 1, label: 'Don\'t Prefer' },
    { value: 2, label: 'Intolerant' },
    { value: 3, label: 'Severe\nAllergy' },
  ];

  const choice = (v) => {
    const halfRatio = sliderRatio / 2;
    // console.log(halfRatio);
    let i = 0;
    for (; ;) {
      if ((v < sliderRatio) || (v <= 0)) {
        if (v >= halfRatio) {
          i++;
        }
        break;
      }
      v = v - sliderRatio;
      i++;
    }
    let value = sliderRatio * i;

    //Move the slider
    value = value + (state.adjustSign * 0.000001);//enforce UI update
    let sign = 1;
    if (state.adjustSign > 0) {
      sign = -1
    } else {
      sign = 1
    }
    // console.log(value);
    setState({ ...state, value: value, adjustSign: sign })
    props.selectState(Math.round(value))
  }

  const styles = StyleSheet.create({
    itemWrapper: {
      marginLeft: 8,
      justifyContent: 'space-between',
      alignSelf: 'stretch',
      flexDirection: 'row',
    },
    item: {
      color: props.color
    },
    sliderContainer: {
      marginTop: 20,
      height: 35,
      display: 'flex',
      justifyContent: 'center',
      borderLeftWidth: 15,
      borderRightWidth: 15,
      borderRadius: 20,
      borderColor: props.color,
    },
    track: {
      height: 35,
      backgroundColor: props.color,
      display: 'flex',
      justifyContent: 'center',
    },
    thumb: {
      width: 20,
      height: 20,
      backgroundColor: props.color,
      borderColor: 'white',
      borderWidth: 7,
      borderRadius: 10,
    }
  });

  return (
    <View>
      <View style={styles.sliderContainer}>

        <Slider
          color={props.color}
          value={state.value}
          minimumValue={0}
          maximumValue={3}
          onSlidingComplete={(value) => choice(value)}
          trackStyle={styles.track}
          thumbStyle={styles.thumb}
          minimumTrackTintColor="#464F7A"
          maximumTrackTintColor="#464F7A"
        />

      </View>
      <View style={styles.itemWrapper}>
        {sliderOptions.map((option) => {
          return <Text key={option.value}>{option.label}</Text>
        })}
      </View>
    </View>
  )
}



export default CustomSlider;