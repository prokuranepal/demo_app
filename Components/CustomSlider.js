import React, { useState } from 'react';
import { StyleSheet, View, Text } from 'react-native';
import Slider from 'react-native-custom-slider';

const CustomSlider = props => {

  const defaultValue = 0
  const maximumValue = 3;
  const length = 4
  const sliderRatio = maximumValue / (length - 1);
  const [state, setState] = useState({
    value: sliderRatio * defaultValue,
    adjustSign: 1
  });
  const sliderOptions = [
    { value: 0, label: 'Eat' },
    { value: 1, label: 'Don\'t Prefer' },
    { value: 2, label: 'Intolerant' },
    { value: 3, label: 'Severe{\n}Allergy' },
  ];

  const choice = (v) => {
    const halfRatio = sliderRatio / 2;
    // console.log(halfRatio);
    let i = 0;
    for (; ;) {
      if ((v < sliderRatio) || (v <= 0)) {
        if (v >= halfRatio) {
          i++;
        }
        break;
      }
      v = v - sliderRatio;
      i++;
    }
    let value = sliderRatio * i;

    //Move the slider
    value = value + (state.adjustSign * 0.000001);//enforce UI update
    let sign = 1;
    if (state.adjustSign > 0) {
      sign = -1
    } else {
      sign = 1
    }
    // console.log(value);
    setState({ ...state, value: value, adjustSign: sign })
    props.selectState(Math.round(value))
  }


  return (
    <View>
      <View style={styles.sliderContainer}>

        <Slider
          value={state.value}
          minimumValue={0}
          maximumValue={3}
          onSlidingComplete={(value) => choice(value)}
          trackStyle={styles.track}
          thumbStyle={styles.thumb}
          minimumTrackTintColor="#464F7A"
          maximumTrackTintColor="#464F7A"
        />

      </View>
      <View style={styles.itemWrapper}>
        {/* {sliderOptions.map((option) => {
          <Text>{option.label}</Text>
        })} */}
        <Text style={styles.item}>Eat</Text>
        <Text style={styles.item}>Don't Prefer</Text>
        <Text style={styles.item}>Intolerant</Text>
        <Text style={styles.item}>Severe{"\n"}Allergy</Text>
      </View>
    </View>
  )
}

const styles = StyleSheet.create({
  itemWrapper: {
    marginLeft: 10,
    justifyContent: 'space-between',
    alignSelf: 'stretch',
    flexDirection: 'row',
  },
  item: {
    color: '#464F7A'
  },
  sliderContainer: {
    marginTop: 20,
    borderLeftWidth: 10,
    borderRightWidth: 15,
    height: 40,
    paddingBottom: 20,
    borderRadius: 20,
    backgroundColor: '#464F7A',
    borderColor: '#464F7A',
  },
  track: {
    
  },
  thumb: {
    width: 20,
    height: 20,
    backgroundColor: '#464F7A',
    borderColor: 'white',
    borderWidth: 5,
    borderRadius: 10,
  }
});

export default CustomSlider;