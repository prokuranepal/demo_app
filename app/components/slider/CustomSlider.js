import React, { useState } from 'react';
import { StyleSheet, View, Text } from 'react-native';

import PropTypes from 'prop-types';

import Slider from './SliderContainer';;

const CustomSlider = props => {

  const maximumValue = props.preferences.length - 1;
  const length = props.preferences.length;
  const sliderRatio = maximumValue / (length - 1);

  const [state, setState] = useState({
    value: props.initialValue,
    adjustSign: 1
  });

  const choice = (v) => {
    const halfRatio = sliderRatio / 2;
    let i = 0;
    for (; ;) {
      if ((v < sliderRatio) || (v <= 0)) {
        if (v >= halfRatio) {
          i++;
        }
        break;
      }
      v = v - sliderRatio;
      i++;
    }
    let value = sliderRatio * i;

    //Move the slider
    value = value + (state.adjustSign * 0.000001);//enforce UI update
    let sign = 1;
    if (state.adjustSign > 0) {
      sign = -1
    } else {
      sign = 1
    }
    // console.log(value, v)
    setState({ ...state, value: value, adjustSign: sign });
    props.selectState(Math.round(value));
  }

  

  return (
    <View style={styles.mainView}>
      <View style={{...styles.sliderContainer,borderColor: props.color}}>

        <Slider
          color={props.color}
          value={state.value}
          minimumValue={0}
          maximumValue={maximumValue}
          onSlidingComplete={(value) => choice(value)}
          trackStyle={{...styles.track,backgroundColor: props.color}}
          thumbStyle={{...styles.thumb,backgroundColor: props.color}}
          minimumTrackTintColor="#464F7A"
          maximumTrackTintColor="#464F7A"
          length={length}
          data-test="sliderComp"
        />

      </View>
      <View style={styles.itemWrapper}>
        {props.preferences.map((option) => {
          return <Text key={option._id} data-test="textComp">{option.sliderText}</Text>
        })}
      </View>
    </View>
  )
};

const styles = StyleSheet.create({
  mainView: {
    flex:1, justifyContent:"center"
  },
  itemWrapper: {
    marginLeft: 8,
    justifyContent: 'space-between',
    alignSelf: 'stretch',
    flexDirection: 'row',
    marginBottom:20,
    marginTop:8
  },
  sliderContainer: {
    marginTop: 20,
    height: 30,
    display: 'flex',
    justifyContent: 'center',
    borderLeftWidth: 15,
    borderRightWidth: 15,
    borderRadius: 16,
  },
  track: {
    height: 30,
    display: 'flex',
    justifyContent: 'center',
  },
  thumb: {
    width: 20,
    height: 20,  
    borderColor: 'white',
    borderWidth: 7,
    borderRadius: 10,
  }
});

CustomSlider.propTypes = {
  preferences: PropTypes.array,
  color: PropTypes.string,
  initialValue: PropTypes.number,
  selectState: PropTypes.func
}


export default CustomSlider;