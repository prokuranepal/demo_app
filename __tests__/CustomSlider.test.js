import React from 'react';
import {View, Text} from 'react-native';
import {configure, shallow} from 'enzyme';
import EnzymeAdapter from 'enzyme-adapter-react-16';
import CustomSlider from '../app/components/slider/CustomSlider';
import Slider from '../app/components/slider/SliderContainer';

import {findByTestAttr} from '../app/utils/testFunction';
// JestHook.mock('expo-font');
configure({adapter: new EnzymeAdapter})

const fn_call = jest.fn()
const dummy_data = {
    selectState: fn_call,
    preferences: [
        {
            _id: 0,
            sliderText: "slide at 0"
        }, {
            _id: 1,
            sliderText: "slide at 1"
        }
    ],
    color: "red",
    initialValue: 1
}

describe('<CustomSlider/>', () => {

    it('should render without issues', () => {
        const component = shallow(<CustomSlider {...dummy_data}/>);
        expect(component.length).toBe(1);
        expect(component).toMatchSnapshot();
    });

    it("Children test in CustomSlider", () => {
        const wrapper = shallow(<CustomSlider {...dummy_data}/>)
        expect(wrapper.find(View)).toHaveLength(3)
        expect(wrapper.find(Text)).toHaveLength(2)
        expect(wrapper.find(Slider)).toHaveLength(1)

    });

    it("Slider Component test in CustomSlider", () => {
        const wrapper = shallow(<CustomSlider {...dummy_data}/>)
        let sliderComp = findByTestAttr(wrapper, "sliderComp");
        expect(sliderComp).toHaveLength(1)
        expect(sliderComp.props().color).toEqual("red")
        expect(sliderComp.props().value).toEqual(1)
        sliderComp
            .props()
            .onSlidingComplete(0)
        wrapper.update()
        expect(fn_call).toHaveBeenCalledTimes(1)
        expect(fn_call).toHaveBeenCalledWith(0)
        sliderComp = findByTestAttr(wrapper, "sliderComp");
        expect(sliderComp.props().value).toEqual(0.000001)
        //   const wrapper2 = shallow(<CustomSlider {...{...dummy_data,
        // visible:true}}>Headerrr</ModalContainer>)   const modalComp2 =
        // findByTestAttr(wrapper2,"sliderComp");
        // expect(modalComp2.props().visible).toEqual(true)
    });
    it("Textr Component test in CustomSlider", () => {
        const wrapper = shallow(<CustomSlider {...dummy_data}/>)
        let textComp = findByTestAttr(wrapper, "textComp");
        expect(textComp).toHaveLength(2)
        expect(textComp.at(0).props().children).toEqual("slide at 0")
        expect(textComp.at(1).props().children).toEqual("slide at 1")


    });

})