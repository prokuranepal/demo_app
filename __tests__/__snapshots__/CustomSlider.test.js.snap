// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomSlider/> should render without issues 1`] = `ShallowWrapper {}`;
